src/libcore/alloc.rs

```diff
@@ -1217,8 +1199,38 @@ pub unsafe trait Alloc {
     /// constraints.
     ///
     /// Always returns `Err` on arithmetic overflow.
-    unsafe fn dealloc_array<T>(&mut self, ptr: NonNull<T>, n: usize) -> Result<(), AllocErr>
-        where Self: Sized
+    unsafe fn dealloc_array<T>(&mut self, ptr: NonNull<T>, n: usize) -> Result<(), Self::Err>
+        where Self: Sized;
+}
+
+#[unstable(feature = "allocator_api", issue = "32838")]
+default unsafe impl<A: AllocHelper<Err = AllocErr>> Alloc for A {
```

Comment:
This seems to cause a "conflicting implementations of trait `core::alloc::Alloc`" error for allocator crates without `#![feature(specialization)]`. The full error is:

```
error[E0119]: conflicting implementations of trait `core::alloc::Alloc` for type `Heap`:
   --> /…/linked-list-allocator/src/lib.rs:136:1
    |
136 | unsafe impl Alloc for Heap {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `core`:
            - impl<A> core::alloc::Alloc for A
              where <A as core::alloc::AllocHelper>::Err == core::alloc::AllocErr, A: core::alloc::AllocHelper, A: core::alloc::Alloc;
```

With the feature added to the crate, it compiles fine.

Reply:
@phil-opp sorry to get back to you so late. Firstly, there is a proposal to remove the offending methods for which I need this `AllocHelper` business in the first place. If they aren't however, I'd hope we could use something lighter than specialization for this. They key property is if the choice if default method can be made without regard to the specific parameter, there is no parametric violation, and thus no interesting "specialization". Haskell has had https://ghc.gitlab.haskell.org/ghc/doc/users_guide/glasgow_exts.html#extension-DefaultSignatures along these lines with no global interactions for ages. We should add this limited form of default impls to Rust, and fix https://github.com/rust-lang/rust/issues/52396 while we are at it.